src/components/structures/login/Login.js

```diff
@@ -285,119 +300,53 @@ module.exports = React.createClass({
     _tryWellKnownDiscovery: async function(serverName) {
         if (!serverName.trim()) {
             // Nothing to discover
-            this.setState({discoveryError: "", discoveredHsUrl: "", discoveredIsUrl: ""});
+            this.setState({discoveryError: "", discoveredHsUrl: "", discoveredIsUrl: "", findingHomeserver: false});
             return;
         }
 
+        this.setState({findingHomeserver: true});
         try {
-            const wellknown = await this._getWellKnownObject(`https://${serverName}/.well-known/matrix/client`);
-            if (!wellknown["m.homeserver"]) {
-                console.error("No m.homeserver key in well-known response");
-                this.setState({discoveryError: _t("Invalid homeserver discovery response")});
-                return;
-            }
-
-            const hsUrl = this._sanitizeWellKnownUrl(wellknown["m.homeserver"]["base_url"]);
-            if (!hsUrl) {
-                console.error("Invalid base_url for m.homeserver");
-                this.setState({discoveryError: _t("Invalid homeserver discovery response")});
-                return;
-            }
-
-            console.log("Verifying homeserver URL: " + hsUrl);
-            const hsVersions = await this._getWellKnownObject(`${hsUrl}/_matrix/client/versions`);
-            if (!hsVersions["versions"]) {
-                console.error("Invalid /versions response");
-                this.setState({discoveryError: _t("Invalid homeserver discovery response")});
-                return;
-            }
-
-            let isUrl = "";
-            if (wellknown["m.identity_server"]) {
-                isUrl = this._sanitizeWellKnownUrl(wellknown["m.identity_server"]["base_url"]);
-                if (!isUrl) {
-                    console.error("Invalid base_url for m.identity_server");
-                    this.setState({discoveryError: _t("Invalid homeserver discovery response")});
-                    return;
-                }
-
-                console.log("Verifying identity server URL: " + isUrl);
-                const isResponse = await this._getWellKnownObject(`${isUrl}/_matrix/identity/api/v1`);
-                if (!isResponse) {
-                    console.error("Invalid /api/v1 response");
-                    this.setState({discoveryError: _t("Invalid homeserver discovery response")});
-                    return;
-                }
+            const discovery = await AutoDiscovery.findClientConfig(serverName);
+            const state = discovery["m.homeserver"].state;
+            if (state !== AutoDiscovery.SUCCESS && state !== AutoDiscovery.PROMPT) {
```

Comment:
Hmm, a bit of a nit, and probably a bit late to bring this up, but I can't help but feeling this is a lot of code and branching for what we're doing here. Maybe I'm missing something, but AutoDiscovery tries to resolve a server name to a hs/is url, or can fail with a message (which can be fatal or not if we want to to make the FAIL_PROMPT distinction, which seems unused atm). If AutoDiscovery.findClientConfig returns a promise that resolved to an object with hsUrl & isUrl fields or rejects with an error message (and fatal field if needed), much of this code could be simplified to just 1 setState call in the try and one in the catch.

Not too important though, so feel free to ignore this.

Reply:
The branching looks a bit bad mostly because I ended up using long variable names, which means a lot of line breaks
for clarity and to avoid making the linter angry. This is only really handling four cases: failure, success, prompt, and something is really broken (invalid status). We end up treating "prompt" as "please use the custom server options" whereas failure of any kind we fall back on preferring to communicate the problem to the user.

For FAIL_PROMPT, we could handle it as a warning (as it generally only happens when the config is wrong) and let the user go through with the custom server options. However, we don't really have a design for communicating a non-fatal error to the user and encourage use of another component. In the interest of trying to get the user to fix their .well-known config, we take the spec's suggestion to fail to heart and communicate the problem to the user.

In an ideal world, the administrator setting up .well-known would test their changes to make sure they work before their users play with it. Theory is that there are going to be few bad configs out there, and most users will easily breeze through the process anyways. 

As for the verboseness: this is an artifact of trying to get all the validation rules communicated and handled in the js-sdk, as desired by the community.